predicators/envs/spot_env.py

```diff
@@ -194,102 +213,149 @@ def build_action(self, state: State, op: STRIPSOperator,
         # Finalize action.
         return Action(action_arr)
 
-    def step(self, action: Action) -> Observation:  # pragma: no cover
+    def reset(self, train_or_test: str, task_idx: int) -> Observation:
+        init_atoms = self._get_initial_nonpercept_atoms()
+        init_obs = self._build_observation(init_atoms)
+        goal = self._generate_task_goal()
+        self._current_task = EnvironmentTask(init_obs, goal)
+        self._current_observation = init_obs
+        return init_obs
+
+    def step(self, action: Action) -> Observation:
         """Override step() because simulate() is not implemented."""
-        state = self._current_observation
-        assert isinstance(state, _PartialPerceptionState)
+        obs = self._current_observation
+        assert isinstance(obs, _SpotObservation)
         assert self.action_space.contains(action.arr)
         # Parse the action into the components needed for a controller.
-        name, objects, params = self.parse_action(state, action)
+        name, objects, params = self.parse_action(action)
         # Execute the controller in the real environment.
-        current_atoms = utils.abstract(state, self.predicates)
-        self._spot_interface.execute(name, current_atoms, objects, params)
-        # Get the part of the new state that is determined based on
-        # continuous feature values.
-        next_state = self._get_continuous_observation()
+        self._spot_interface.execute(name, objects, params)
         # Now update the part of the state that is cheated based on the
         # ground-truth STRIPS operators.
-        next_sim_state_ground_atoms = self._get_next_simulator_state(
-            state, action)
-        if next_sim_state_ground_atoms is None:  # inapplicable action
-            return state.copy()
-        # Combine the two to get the new _PartialPerceptionState.
-        self._current_observation = self._build_partial_perception_state(
-            next_state.data, next_sim_state_ground_atoms)
-        return self._current_observation.copy()
-
-    def _build_partial_perception_state(
-            self, state_data: Dict[Object, Array],
-            ground_atoms: Set[GroundAtom]) -> _PartialPerceptionState:
-        """Helper for building a new _PartialPerceptionState().
-
-        This is an environment method because the predicates stored in
-        the simulator_state may vary per environment.
+        next_nonpercept = self._get_next_nonpercept_atoms(obs, action)
+        self._current_observation = self._build_observation(next_nonpercept)
+        return self._current_observation
+
+    def get_observation(self) -> Observation:
+        return self._current_observation
+
+    def goal_reached(self) -> bool:
+        # We need to implement this! But we're just watching it work for now.
+        return False
```

Comment:
I really think a reasonable thing to do here is to ask a user/human/oracle for input on this. In general, i feel like it might be hard to classify goals purely from observation (since they so often require predicates and a notion of `state`), so it might be good to have this human-asking functionality as a general thing